File: agent/dotnet/src/Perper.Protocol/CallData.cs
using System;
using System.Collections;
using System.Linq;
using System.Collections.Generic;
using System.Threading.Tasks;
using Apache.Ignite.Core;
using Apache.Ignite.Core.Binary;
using Apache.Ignite.Core.Cache.Affinity;
using Apache.Ignite.Core.Client;

namespace Perper.Protocol
{
    static class CallData
    {
        public static IBinaryObject CreateCallData<TParams>(
            IBinary binary,
            string instance,
            string agent,
            string @delegate,
            string callerAgent,
            string caller,
            bool localToData,
            TParams parameters)
        {
            var builder = binary.GetBuilder($"CallData_{agent}_{@delegate}");

            builder.SetField("agent", agent);
            builder.SetField("instance", instance);
            builder.SetField("delegate", @delegate);
            builder.SetField("callerAgent", callerAgent);
            builder.SetField("caller", caller);
            builder.SetField("finished", false);
            builder.SetField("localToData", localToData);
            builder.SetField("parameters", parameters);

            return builder.Build();
        }

        public static IBinaryObject SetCallDataResult<TResult>(
            IBinaryObject existingCallData,
            TResult result)
        {
            var builder = existingCallData.ToBuilder();

            builder.SetField("finished", true);
            builder.SetField("result", result);

            return builder.Build();
        }

        public static IBinaryObject SetCallDataResult(IBinaryObject existingCallData) // Void result
        {
            var builder = existingCallData.ToBuilder();

            builder.SetField("finished", true);

            return builder.Build();
        }

        public static IBinaryObject SetCallDataError(
            IBinaryObject existingCallData,
            string error)
        {
            var builder = existingCallData.ToBuilder();

            builder.SetField("finished", true);
            builder.SetField("error", error);

            return builder.Build();
        }
    }
}

File: agent/python/perper/call_data.py
from collections import OrderedDict
from pyignite import GenericObjectMeta
from pyignite.datatypes import String, BoolObject

def create_call_data(instance, agent, delegate, caller_agent, caller, local_to_data, parameters, parameters_type):
    class CallData(metaclass=GenericObjectMeta, type_name=f"CallData_{agent}_{delegate}", schema=OrderedDict([
        ('instance', String),
        ('agent', String),
        ('delegate', String),
        ('callerAgent', String),
        ('caller', String),
        ('finished', BoolObject),
        ('localToData', BoolObject),
        ('parameters', parameters_type),
        #('result', result_type),
    ])):
        pass

    return CallData(
        instance=instance,
        agent=agent,
        delegate=delegate,
        callerAgent=caller_agent,
        caller=caller,
        finished=False,
        localToData=local_to_data,
        parameters=parameters
    )

def set_call_data_result(call_data, result, result_type):

    schema = call_data.schema
    schema['result'] = result_type

    class CallData(metaclass=GenericObjectMeta, type_name=call_data._type_name, schema=schema):
        pass

    return CallData(
        finished=True,
        result=result,
        **{key: getattr(call_data, key) for key in schema.keys() - {'finished', 'result'}}
    )